geps/gep-1911.md

```diff
@@ -0,0 +1,190 @@
+  //   - port: 80
+  //     http1: {}          # We also listen to http1 on port 80
+  //
+  Ports []BackendPort `json:"ports,omitempty"`
+}
+
+type BackendPort struct {
+  // Number or name of the port who's protocol is being defined
+  Port intstr.IntOrString
+
+  PortProtocol `json:",inline"`
+}
+
+type PortProtocol struct {
+  HTTP2     *HTTP2PortProtocol    `json:"http2"`
+  HTTP1     *HTTP1PortProtocol    `json:"http1"`
+  GRPC      *GRPCPortProtocol     `json:"grpc"`
+  WebSocket *WebsocketPortProtocol`json:"websocket"`
+}
+
+type HTTP2PortProtocol struct {
+  Cleartext bool // h2c when set to true
+
+  // Future knobs can go here
+}
+
+type HTTPPortProtocol struct {
+  // Future knobs can go here
+}
+type GRPCPortProtocol struct {
+  // Future knobs can go here
+}
+type WebsocketPortProtocol struct {
+  // Future knobs can go here
+}
+```
+
+### YAML API
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1alpha1
+kind: BackendProtocol
+metadata:
+  name: some-service
+  namespace: some-namespace
+spec:
+  ports:
+  - port: 80
+    http2: 
+      encrypted: false # enable h2c
+  - port: 80
+    http1: {}          # We also listen to http1 on port 80
+  - port: 9090
+    grpc: {}
+  - port: 8080
+    websocket: {}      # enable websocket support on this port
+  - port: 7070
+    http1: {}
+  targetRef:
+    group: "" # corev1
+    kind: Service
+    name: some-service
+```
+
+### Default Protocols
+
+If a backend protocol isn't specified an implementation MAY infer the backend protocol through it's own means.
+
+### Incompatible Protocols
+
+If a Route attached to a Gateway is not able to support sending traffic to the backend using a specific protocol 
+then it must set it's condition to `Accepted` `False`.
+
+## Alternatives
+
+### Meta-resource for each protocol
+
+Rather than bundle protocol details into a single resource an alternative would be to create distinct meta resources.
+ie. `HTTP2Backend`, `GPRCBackend`, `WebsocketBackend`.
+
+Although this has the advantage of seperate it doesn't allow
+
+### Adding properties on Gateway Route Objects
+
+From [GEP-1282](geps/gep-1282.md#tldr):
+> some types of configuration requested by users are more about defining functionality that describes capabilities of the backend more than the route you take to get to the backend.
+
+Backend protocol is specifying capabilities. This configuration is less about routing.
+
+### Extending Kubernetes Service
+
+This is considered untenable due to the 'the turnaround time for those changes can be years.' ([ref-1282](geps/gep-1282.md#non-goals):
+
+### AppProtocol
+
+From [GEP-1282](geps/gep-1282.md#non-goals):
+> v1.Service’s appProtocol field is not fit for purpose, because it is defined as accepting values either from the IANA Service Name registry, or domain-prefixed values and we need more flexibility than that. 
+
+See further discussion [here](https://github.com/kubernetes-sigs/gateway-api/discussions/1244#discussioncomment-3151197)
```

Comment:
I am not sure this is true since https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/3726-standard-application-protocols#summary?

There was talk of allowing multiple appProtocols as well - I don't think it landed but I don't think its in-plausible

Reply:
Yep I think multiple app protocols is inevitable - help wanted though. @LiorLieberman has been helping move AppProtocols forward upstream.

Reply:
this would be needed for all backend types including ServiceImport
https://github.com/kubernetes-sigs/mcs-api/blob/11eea5e71266160cc900f319b3f080c24dbea4f1/pkg/apis/v1alpha1/serviceimport.go#L99

Reply:
> There was talk of allowing multiple appProtocols as well - I don't think it landed but I don't think its in-plausible

Is that this https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1435-mixed-protocol-lb ? It's GA in K8s 1.26. Do we know why the KEP is scoped to type=LoadBalancer?

> I am not sure this is true since https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/3726-standard-application-protocols#summary?

The examples are confusing - it suggests `HTTP2` has two different meanings (one is http2 over tls in GKE and the other is h2c with Istio). Not a fan of that.

I think we need to decided do we want the ability to customize the backend in any way - https://github.com/kubernetes-sigs/gateway-api/pull/1979#issuecomment-1526585087

Reply:
>The examples are confusing - it suggests HTTP2 has two different meanings (one is http2 over tls in GKE and the other is h2c with Istio). Not a fan of that.

To clarify, that was trying to note that the same values were being interpreted different ways by different implementations. It was not an example of something to do, more something this KEP was trying to solve.

Reply:
I shouldn't skim KEPs during dinner ;)

Reply:
One observation I have about appProtocol (or `lists of protocols` in general) is how can a backend _disable_ a protocol.

ie. disable websockets & http

Reply:
Added some comments in the GEP from this discussion

Following up on[KEP-1435 - Mixed Protocol](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1435-mixed-protocol-lb) - it only applies to the `protocol` field. Thus trying to specify a K8s Service with multiple `appProtocol` to a single port won't work

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
  labels:
    app.kubernetes.io/name: proxy
spec:
  containers:
  - name: nginx
    image: nginx:stable
    ports:
      - containerPort: 80
        name: http-web-svc
---
apiVersion: v1
kind: Service
metadata:
  name: nginx-service
spec:
  selector:
    app.kubernetes.io/name: proxy
  ports:
  - name: http
    protocol: TCP
    appProtocol: http
    port: 80
    targetPort: http-web-svc
  - name: h2c
    appProtocol: h2c
    protocol: TCP
    port: 80
    targetPort: http-web-svc
```

Reply:
Switched the GEP to consider `appProtocol` resolving - please review